apps/api/src/user/service/user.service.ts

```diff
@@ -60,6 +127,95 @@ export class UserService {
     })
   }
 
+  async validateEmailChangeOtp(user: User, otp: string): Promise<User> {
+    const userEmailChange = await this.prisma.userEmailChange.findUnique({
+      where: {
+        userId_otp: {
+          userId: user.id,
+          otp: otp
+        },
+        expiresOn: {
+          gt: new Date()
+        }
+      }
+    })
+
+    if (!userEmailChange) {
+      this.log.log(`OTP expired or invalid`)
+      throw new UnauthorizedException('Invalid or expired OTP')
+    }
+
+    const deleteEmailChangeRecord = this.prisma.userEmailChange.delete({
+      where: {
+        userId_otp: {
+          userId: user.id,
+          otp: otp
+        }
+      }
+    })
+
+    const deleteOtp = this.prisma.otp.delete({
+      where: {
+        userId: user.id,
+        code: otp
+      }
+    })
+
+    const updateUserOp = this.prisma.user.update({
+      where: {
+        id: user.id
+      },
+      data: {
+        email: userEmailChange.newEmail,
+        authProvider: AuthProvider.EMAIL_OTP
+      }
+    })
+
+    this.log.log(
+      `Changing email to ${userEmailChange.newEmail} for user ${user.id}`
+    )
+    const results = await this.prisma.$transaction([
+      deleteEmailChangeRecord,
+      deleteOtp,
+      updateUserOp
+    ])
+
+    return results[2]
+  }
+
+  async resendEmailChangeOtp(user: User) {
+    const oldOtp = await this.prisma.otp.findUnique({
+      where: {
+        userId: user.id
+      }
+    })
+
+    if (!oldOtp) {
+      throw new ConflictException(`No previous OTP exists for user ${user.id}`)
+    }
+
+    const newOtp = await generateOtp(user.email, user.id, this.prisma)
+
+    const newUserEmailChange = await this.prisma.userEmailChange.update({
```

Comment:
As per the new functionality, I think you can get rid of this block.

Reply:
Removed